Convert dropdowns and related inputs to pf4

sg00dwin opened the pull request.

spadgett commented:
cc @christianvogt

spadgett commented:
cc @priley86

rhamilto commented on frontend/public/components/_cluster-service-instance.scss:
```diff
@@ -3,8 +3,7 @@
 .co-create-service-instance__namespace {
-  .btn-dropdown,
-  .dropdown-menu {
+  .pf-c-dropdown {
```
Nit:  remove nesting.

rhamilto commented on frontend/public/components/_deploy-image.scss:
```diff
@@ -5,8 +5,7 @@
   &__namespace {
-    .btn-dropdown,
-    .dropdown-menu {
+    .pf-c-dropdown {
```
Nit:  remove nesting.

rhamilto commented on frontend/public/components/_dropdown.scss:
```diff
@@ -1,7 +1,6 @@
-$color-bookmarker--hover: #AAA;
-$color-bookmarker: #DDD;
-$dropdown-link-hover-border: 1px solid #bee1f4;
+$color-bookmarker--hover: #999;
```
Nit:  here and next line:  use pf color vars?  Maybe `$color-pf-black-500` and `$color-pf-black-400`?

rhamilto commented on frontend/public/components/_dropdown.scss:
```diff
@@ -60,55 +59,43 @@ $dropdown-link-hover-border: 1px solid #bee1f4;
-  }
-
+// Custom dropdown menu rules for namespace-selector & resource-item that don't use the default pf4 styles
+.co-namespace-selector .dropdown-menu__autocomplete-filter {
```
Nit:  here and next grouping:  alpha

rhamilto commented on frontend/public/components/_dropdown.scss:
```diff
@@ -60,55 +59,43 @@ $dropdown-link-hover-border: 1px solid #bee1f4;
       outline: none;
     }
+    &.active {
+      a {
```
Nit:  remove nesting

rhamilto commented on frontend/public/components/_dropdown.scss:
```diff
@@ -193,15 +181,17 @@ $dropdown-link-hover-border: 1px solid #bee1f4;
   max-width: 100%;
   min-width: 210px;
   overflow-y: auto;
+  top: 41px; // namespace selector bar height - 1px (bottom border)
```
Wonder if it'd be better to use top margin instead to make this a little less brittle?

rhamilto commented on frontend/public/components/_dropdown.scss:
```diff
@@ -213,62 +203,33 @@ $dropdown-link-hover-border: 1px solid #bee1f4;
         padding-right: 5px;
         padding-left: 12px;
         color: $color-bookmarker;
+        flex-basis: 0;
```
Use `flex` instead?

rhamilto commented on frontend/public/components/_dropdown.scss:
```diff
@@ -213,62 +203,33 @@ $dropdown-link-hover-border: 1px solid #bee1f4;
-      &:before {
-        bottom: 0;
-      }
       &:hover {
```
Nit:  remove nesting

rhamilto commented on frontend/public/components/_dropdown.scss:
```diff
@@ -297,19 +274,15 @@ $dropdown-link-hover-border: 1px solid #bee1f4;
   font-style: italic;
 }
 
 .btn-dropdown__content-wrap {
```
Nit:  alpha

rhamilto commented on frontend/public/components/_source-to-image.scss:
```diff
@@ -1,6 +1,5 @@
 .co-source-to-image-form {
-  .btn-dropdown,
-  .dropdown-menu {
+  .pf-c-dropdown {
```
Nit:  remove nesting

rhamilto commented on frontend/public/components/cluster-settings/_cluster-settings.scss:
```diff
@@ -1,7 +1,6 @@
 .cluster-update-modal__dropdown {
-  .btn-dropdown,
-  .dropdown-menu {
+  .pf-c-dropdown {
```
Nit:  remove nesting

rhamilto commented on frontend/public/style/_overrides.scss:
```diff
@@ -9,147 +9,6 @@ $pf-4-nav-bar-height: 76px; // Height of the PatternFly 4 masthead
   cursor: auto;
 }
 
-tags-input:focus,
```
Moving these lines out of this file breaks our convention of putting overrides of third-party css in a single file and results in us having overrides throughout.  I'm not opposed to that, but do favor consistency.

sg00dwin replied:
Turns out we are not overriding existing 'react-tagsinput`

rhamilto commented on frontend/public/vendor.scss:
```diff
@@ -56,7 +56,14 @@
 @import '~patternfly-react-extensions/dist/sass/filter-side-panel';
 @import '~patternfly-react-extensions/dist/sass/properties-side-panel';
 @import '~patternfly-react-extensions/dist/sass/vertical-tabs';
+
```
You can remove all the changes in this file as they are unnecessary/undesirable.

rhamilto commented on frontend/public/style/_overrides.scss:
```diff
@@ -509,6 +352,18 @@ kbd {
   --pf-c-page__sidebar-body--PaddingBottom: 0;
 }
 
+// specificity targeting form elements to override --pf-global--FontSize--md
```
👍

priley86 commented:
these dropdowns look a lot more consistent to me now, nice work @sg00dwin!

Just noting a few things after glancing at this and starting #1629 previously (not sure if they were already noted):
 
* Events and Search dropdown items could probably use some design adjustment. cc: @beanh66 
<img width="429" alt="Screen Shot 2019-06-19 at 10 03 44 AM" src="https://user-images.githubusercontent.com/4237045/59772648-2b33ef00-927a-11e9-8d04-0468f72ba831.png">
<img width="527" alt="Screen Shot 2019-06-19 at 10 01 16 AM" src="https://user-images.githubusercontent.com/4237045/59772657-2ff8a300-927a-11e9-8cc8-a79ba0054082.png">

* The CSS override block here can ideally be removed completely after this PR. I added this temporarily while retroverting the PF4 Kebab in #1465 
https://github.com/openshift/console/blob/master/frontend/public/style/_overrides.scss#L603-L634

sg00dwin commented:
> these dropdowns look a lot more consistent to me now, nice work @sg00dwin!
> 
> Just noting a few things after glancing at this and starting #1629 previously (not sure if they were already noted):
> 
>     * Events and Search dropdown items could probably use some design adjustment. cc: @beanh66
> 
> 
> <img alt="Screen Shot 2019-06-19 at 10 03 44 AM" width="429" src="https://user-images.githubusercontent.com/4237045/59772648-2b33ef00-927a-11e9-8d04-0468f72ba831.png">
> 
> <img alt="Screen Shot 2019-06-19 at 10 01 16 AM" width="527" src="https://user-images.githubusercontent.com/4237045/59772657-2ff8a300-927a-11e9-8cc8-a79ba0054082.png">
> 
>     * The CSS override block here can ideally be removed completely after this PR. I added this temporarily while retroverting the PF4 Kebab in #1465
>       https://github.com/openshift/console/blob/master/frontend/public/style/_overrides.scss#L603-L634

thanks @priley86 those dropdowns where missing a rule to control min-width. And I'll address those kebab overrides.

sg00dwin commented:
@rhamilto Addressed comments ptal

Corrected Events and Search dropdown-menu widths
Right alignment of kebab dropdown-menu with list table
Alphabetize css rules in `_overrides.scss`

<img width="228" alt="Screen Shot 2019-06-19 at 3 47 46 PM" src="https://user-images.githubusercontent.com/1874151/59799438-5803f880-92b2-11e9-9a5e-0cc6a8d13649.png">
<img width="560" alt="Screen Shot 2019-06-19 at 4 47 26 PM" src="https://user-images.githubusercontent.com/1874151/59799447-5b977f80-92b2-11e9-9461-25df8d17272f.png">
<img width="477" alt="Screen Shot 2019-06-19 at 4 47 38 PM" src="https://user-images.githubusercontent.com/1874151/59799448-5b977f80-92b2-11e9-9031-582111b66721.png">

rhamilto commented on frontend/public/vendor.scss:
```diff
@@ -56,7 +56,10 @@
 @import '~patternfly-react-extensions/dist/sass/filter-side-panel';
 @import '~patternfly-react-extensions/dist/sass/properties-side-panel';
 @import '~patternfly-react-extensions/dist/sass/vertical-tabs';
+
```
Remove extra lines on 59, 64, 65

rhamilto commented:
/retest

sg00dwin commented:
/retest

sg00dwin commented:
/retest

rhamilto commented:
/lgtm

christianvogt commented:
The recent additional of the dev console added a new component in the secondary masthead alongside the namespace selector, sorry ;/
![image](https://user-images.githubusercontent.com/14068621/59874432-30716680-936c-11e9-8bc0-45d906354111.png)

Looks like this component `packages/dev-console/src/components/dropdown/ApplicationSelector.tsx` needs `buttonClassName="pf-m-plain"`

spadgett commented:
/hold

sg00dwin commented:
/retest

spadgett commented:
/hold cancel

spadgett commented on frontend/public/style/_overrides.scss:
```diff
@@ -358,6 +147,23 @@ kbd {
   }
 }
 
+// specificity targeting form elements to override --pf-global--FontSize--md
```
What impact does this have if we change this globally to `$font-size-base` as we transition to PF4? Does it break the nav/masthead? The dev console team is looking at bringing in some other PF4 components and will have similar font-size issues.

cc @rhamilto @christianvogt

christianvogt replied:
It would be nice if we can override `--pf-global--FontSize--md` and set it to 14px so that every new PF4 component we utilize doesn't need a separate font tweak.

sg00dwin replied:
@spadgett Setting `--pf-global--FontSize--md` to `$font-size-base: 14px` globally still requires a few targeted overrides. Can we make those global font-size changes as a follow-on and go forward with merging this pf, so the font related changes can be easier tracked?

spadgett replied:
@sg00dwin I was going to suggest a separate PR, yeah. Then we can get that merged separately to unblock the dev console team. It looks like you have some test failures to work through still in this one.

beanh66 commented:
@sg00dwin LGTM! 

@lizsurette @serenamarie125 FYI

lizsurette commented:
Very nice!

FYI @matthewcarleton

sg00dwin commented:
/retest

matthewcarleton commented:
looks great!

rhamilto commented:
/retest

rhamilto commented:
/lgtm

spadgett commented on frontend/integration-tests/views/namespace.view.ts:
```diff
@@ -2,4 +2,4 @@ import { $, $$ } from 'protractor';
 export const namespaceBar = $('.co-namespace-bar');
 export const namespaceSelector = $('.co-namespace-selector');
-export const selectedNamespace = $$('.co-namespace-selector .dropdown .btn-link .btn-link__title').first();
+export const selectedNamespace = $$('.co-namespace-selector .pf-c-dropdown .pf-m-plain .btn-link__title').first();
```
FYI, This will conflict with #1542

spadgett commented on frontend/integration-tests/tests/crud.scenario.ts:
```diff
@@ -272,7 +272,7 @@ describe('Kubernetes resource CRUD operations', () => {
       await crudView.resourceRows.$$('.co-kebab__button').click();
       await browser.wait(until.elementToBeClickable($('.co-kebab__dropdown')));
-      await element(by.cssContainingText('.co-kebab__dropdown a', 'View Instances')).click();
+      await element(by.cssContainingText('.co-kebab__dropdown button', 'View Instances')).click();
```
Prefer this helper for kebab actions:

https://github.com/openshift/console/blob/master/frontend/integration-tests/views/crud.view.ts#L64-L68

rhamilto commented:
/lgtm

rhamilto commented:
/retest
